### ISSUE TITLE
Unable to use google.rpc.Status

### ISSUE THREAD

**Issue by zvbuhl** [gogo/protobuf#372]
In our code at https://github.com/zvelo/msg/ we're using `google.rpc.Status` in some of our own protobufs.  Unfortunately, this results in errors of the following form when building:

```
./dataset.pb.go:270:16: this.Error.Equal undefined (type
    *"zvelo.io/msg/vendor/google.golang.org/genproto/googleapis/rpc/status".Status
    has no field or method Equal)
./dataset.pb.go:308:16: this.Error.Equal undefined (type
    *"zvelo.io/msg/vendor/google.golang.org/genproto/googleapis/rpc/status".Status
    has no field or method Equal)
```

Our team are temporarily working around the issue by forking our own copies of `google.golang.org/genproto` & `github.com/golang/protobuf`, but are curious if modifying our gogo generator somehow would be a better solution. Any thoughts?

**Comment by awalterschulze** [comment]
If you want we could create an genproto "clone" for gogoprotobuf at github.com/gogo/googleapis/, where we generate all the googleapis types using gogoprotobuf with marshalers, equalers, etc.
Would that help?

**Comment by zvbuhl** [comment]
It'd certainly make sense to have it cloned once for all gogoprotobuf users rather than different gogoprotobuf users making our own clones, so it'd help.

Is there any chance of changing gogoprotobuf so that it's un-needed, or is that too great an architectural change?

**Comment by awalterschulze** [comment]
It is semi possible to make this change. So firstly all plugins would have
to be modified to check if the method is there or not, which is not really
possible without incurring a runtime cost. And then that plugin should also
generate the equal etc method in place.

This is a problem I solved in my other code generator goderive, but it does
not generate marshal and unmarshal yet, only equal, compare, gostring, etc.

I think a gogo fork of the types would by far be the easiest thing to do.
Since golang protobuf maintains one, I guess we should too.

What do you think?

**Comment by zvbuhl** [comment]
That's a tough call, but in general I think expedient-and-simple-first, complex-and-complete-second is a good approach to take. Granted, the road to Hell is paved with good intentions:-)

**Comment by kyessenov** [comment]
You could take https://github.com/istio/gogo-genproto/tree/master/googleapis for expediency. Eventually, googleapis might get an official gogo support.

**Comment by awalterschulze** [comment]
Would you like to move that to gogo/googleapis and help me support it?

**Comment by kyessenov** [comment]
I'm certainly open to that. What we have is a short term solution, and it does not really belong to istio. I think googleapis packages google/api and google/rpc are almost mandatory for using gRPC with gogo/proto3. We started adding other third-party protos (opencensus and prometheus). Do you want those, too?

I'd like to hear from @wora and @dougreid about supporting gogo/googleapis.

**Comment by douglas-reid** [comment]
I'd be onboard to move some, if not all, of istio/gogo-genproto to gogo/googleapis . For things like `google.rpc.Status`, I expect very little in the way of required maintenance.

For the other bits, I think that is a tougher call. It would be nice to have some common place to hang gogo-generated protos for repos that don't want / care to commit them. We'd probably want that to be more of a catch-all than gogo/googleapis.

What is the proposed strategy for the generation in istio/gogo-genprot?  Base gogo? gogofaster? gogoslick?

**Comment by awalterschulze** [comment]
Let us decide what methods we want generated and then see which binary
gives us that if any. We can make another one if required.

I did not know there were so many others, but why not have them all in one
common place if more people are using them.

I also think maintenance will be extremely minor. But every now and then
the originating protos gets updated and then we should update the generated
code as well.

**Comment by wora** [comment]
I think we can clone google.rpc, google.type, and a couple of protos in
google.api. They should require very little maintenance, a couple of
changes per year, mostly comments. We can work out long term solution later.

**Comment by kyessenov** [comment]
What's the plan then? @awalterschulze if you create a repo "genproto" under your org, we can start contributing there what we have already in istio/gogo-genproto, and then swap our consumers as well.
Would need some help with setting up CI/CD.

**Comment by awalterschulze** [comment]
I think we should create a separate repo for each set of protos. One for
googleapis, one for prometheus, etc
That way people can minimize their dependencies.

For Ci we could use travis, like I am already using for gogoprotobuf.

So what methods do we want.
I think at least marshal, unmarshal, size, equal, populate.
What do you think?

**Comment by kyessenov** [comment]
Let's start with googleapis. We use gogoslick for googleapis, and that works fine.

**Comment by awalterschulze** [comment]
Sorry for the delay.

gogoslick includes setting everything to nullable=false
https://github.com/gogo/protobuf/blob/master/protoc-gen-gogoslick/main.go#L45
which might not be what everyone wants.
But ok that is only for proto2 and these googleapis are probably all proto3.

It also removes XXX_unrecognized which is something that will be back in proto3 soon
https://github.com/gogo/protobuf/blob/master/protoc-gen-gogoslick/main.go#L46
Here is an example
https://github.com/golang/protobuf/blob/dev/proto/proto3_proto/proto3.pb.go#L83
So I think we should at least discuss whether we want to keep it removed or not?

gogoslick also does some things with prefixes
https://github.com/gogo/protobuf/blob/master/protoc-gen-gogoslick/main.go#L48-L50
Here we should be sure we want these modifications on enums?

**Comment by kyessenov** [comment]
Yes, googleapis is entirely proto3 so nullability is not a problem. I'm not sure about XXX_unrecognized problem; perhaps, @douglas-reid can advise here, since he's been using the generated files in practice.

**Comment by awalterschulze** [comment]
I just remembered that I have actually built a protoc-gen command specifically for the well known types.  This might also be perfect for this use case.  What do you think?
https://github.com/gogo/protobuf/blob/master/protoc-gen-gogotypes/main.go

We can then also add special cases, if it is needed.  For example timestamp and duration requires that their int64 values be in a certain range for testing purposes.  So then the populate generation is turned off in favour of a manually written one.

What do you think?

**Comment by douglas-reid** [comment]
The istio (mixer) codebase uses the options for the enums to make the code less verbose. I don't know that we care overly much about that bit of sugar -- and, when switching SoT for the generated code, can be easily updated. So, I don't know that I would force that on everyone.

Istio (mixer) isn't doing anything that requires preserving or accessing unrecognized fields at the moment. Including them should not affect anything current, and doesn't seem overly problematic. Perhaps we should err on the side of caution?

**Comment by wora** [comment]
In 6 months, Go protobuf will preserve unknown fields because of proto3 spec change. It was announced at golang-nuts@googlegroups.com.

**Comment by awalterschulze** [comment]
So keeping the unrecognized fields, would be erring on the side of caution in my opinion.
Here is the more exact reference
https://groups.google.com/forum/#!topic/golang-nuts/F5xFHTfwRnY
But keeping unknown fields is also more expensive, since it requires a copy and allocation when unmarshaling.  But it should be rare to have unknown values there, so the side of caution wins again.
But it is also an extra pointer for the garbage collector to check, but maybe thats not the worst thing, eh.

@douglas-reid sounds like you are saying you like the proposed command
https://github.com/gogo/protobuf/blob/master/protoc-gen-gogotypes/main.go
Or can you object to specific lines and make it clearer to me?

**Comment by douglas-reid** [comment]
@awalterschulze Yes. Sorry, I should have been more clear. Your proposal sounds reasonable to me.

**Comment by awalterschulze** [comment]
Ok cool :)
Any objections of suggestions from anyone else?

**Comment by awalterschulze** [comment]
Here is a start
https://github.com/gogo/googleapis

**Comment by awalterschulze** [comment]
1. Who is up for adding more protos from googleapis and helping to maintain the repo?
First look at the Makefile before answering please.
The Makefile should be used to update protos as well as regenerating the go code.
I also added a special command to the repo, since compare methods are not properly supported for oneofs.

2. Any objections, suggested modifications or approvals?

3. Sorry @kyessenov for forgetting to look at your repo again.

**Comment by wora** [comment]
The recommended proto3 directory structure is similar to Java, a fully hierarchical packages. Therefore, we could use a single repo to host proto3 files from many companies. Do we want one repo per company? How do we manage license and other issues?

Having the gugo org to host the proto3 files and the generated Go files sound good overall.

**Comment by awalterschulze** [comment]
I think a separate repo per company. Just because the googleapis one
already contains so many.

And I think it makes it easier for people to manage their dependencies and
keep them to a minimum.

**Comment by douglas-reid** [comment]
@awalterschulze i don't mind volunteering to help maintain the repo.

**Comment by awalterschulze** [comment]
Thanks that sounds great.

What do you think about making a pull request and adding some protobufs to
googleapis. That way we get to see if we can work together?

**Comment by douglas-reid** [comment]
sounds good. might be a few days, but i'll put a PR out ;). any targets anyone is looking for to start?

**Comment by awalterschulze** [comment]
No time limit. Take your time.

Also reviews of the googleapis repo are welcome.

**Comment by wora** [comment]
Walter, I have a quick question. The gogo/googleapis uses `google.api` as
Go package name for proto package `google.api`. Is it a legal Go package
name?

If possible, do you mind to add `google.type` package to the gogo
repository? Thanks.

**Comment by awalterschulze** [comment]
The generated go package name is `api`
https://github.com/gogo/googleapis/blob/master/google/api/annotations.pb.go#L4
So that works out fine.

If you mean these protos
https://github.com/googleapis/googleapis/tree/master/google/type
Then that would still fit into the github.com/gogo/googleapis repo under the google/type folder.
Would you mind making a pull request?

**Comment by wora** [comment]
Sorry. You are correct. I was confused between go and proto files. I will
do the pull request next week.

**Comment by awalterschulze** [comment]
Awesome looking forward to it :)
I would do it myself, but I am busy trying to merge golang/protobuf's table marshaling into gogo/protobuf.  It is a big pain.
